src/engine.rs

```diff
@@ -300,33 +309,51 @@ impl Reedline {
     }
 
     fn previous_history(&mut self) {
-        if self.history.history_prefix.is_none() {
-            let buffer = self.line_buffer.get_buffer();
-            self.history.history_prefix = Some(buffer.to_owned());
+        if self.input_mode != InputMode::HistoryTraversal {
+            self.input_mode = InputMode::HistoryTraversal;
         }
 
-        if let Some(history_entry) = self.history.go_back_with_prefix() {
-            let new_buffer = history_entry.to_string();
-            self.set_buffer(new_buffer);
-            self.move_to_end();
+        match (self.line_buffer.is_empty(), self.history.get_navigation()) {
+            (true, HistoryNavigationQuery::Normal) => {}
+            (true, _) => {
+                self.history.set_navigation(HistoryNavigationQuery::Normal);
+            }
+            (false, HistoryNavigationQuery::PrefixSearch(_)) => {}
+            (false, _) => {
+                let buffer = self.insertion_line().to_string();
+                self.history
+                    .set_navigation(HistoryNavigationQuery::PrefixSearch(buffer));
+            }
         }
+
+        self.history.back();
     }
 
     fn next_history(&mut self) {
-        if self.history.history_prefix.is_none() {
-            let buffer = self.line_buffer.get_buffer();
-            self.history.history_prefix = Some(buffer.to_owned());
+        if self.input_mode != InputMode::HistoryTraversal {
+            self.input_mode = InputMode::HistoryTraversal;
         }
 
-        if let Some(history_entry) = self.history.go_forward_with_prefix() {
-            let new_buffer = history_entry.to_string();
-            self.set_buffer(new_buffer);
-            self.move_to_end();
+        match (self.line_buffer.is_empty(), self.history.get_navigation()) {
+            (true, HistoryNavigationQuery::Normal) => {}
+            (true, _) => {
+                self.history.set_navigation(HistoryNavigationQuery::Normal);
+            }
+            (false, HistoryNavigationQuery::PrefixSearch(_)) => {}
+            (false, _) => {
+                let buffer = self.insertion_line().to_string();
+                self.history
+                    .set_navigation(HistoryNavigationQuery::PrefixSearch(buffer));
+            }
         }
+
+        self.history.forward();
     }
```

Comment:
Any way to avoid this duplication?

Reply:
You can move line 342 up:

```rust
            (true, HistoryNavigationQuery::Normal) => {}
            (false, HistoryNavigationQuery::PrefixSearch(_)) => {}
            _ => {
                let buffer = self.insertion_line().to_string();
                self.history
                    .set_navigation(HistoryNavigationQuery::PrefixSearch(buffer));
            }
```

Reply:
In one case I am setting navigation to Normal and in one to `PrefixSearch` so unfortunately I can't really do this.

Reply:
https://github.com/jonathandturner/reedline/pull/92/commits/2a77ebb8c59089046c1276a73bf0a76e9349bcc1 Extracts the command part as an abstraction into a function.